filebeat/crawler/prospector_log.go

```diff
@@ -41,39 +32,49 @@ func (p *ProspectorLog) Init() {
 
 	logp.Info("Load previous states from registry into memory")
 
+	// Load the initial state from the registry
 	for path, fileinfo := range p.getFiles() {
 
 		// Check for each path found, if there is a previous state
 		offset := p.Prospector.registrar.fetchState(path, fileinfo)
 
 		// Offset found -> skip to previous state
 		if offset > 0 {
-			fs := input.NewFileStat(fileinfo, 0)
-			fs.Skip(offset)
-			p.harvesterStats[path] = *fs
+			state := input.NewFileState(fileinfo, path)
+			state.Offset = offset
+			// Make sure new harvester is started for all states
+			state.Finished = true
+			// Prospector must update all states as it has to detect also file rotation
+			p.Prospector.updateState(state)
 		}
 	}
 
-	logp.Info("Previous states loaded: %v", len(p.harvesterStats))
+	logp.Info("Previous states loaded: %v", len(p.Prospector.harvesterStates))
 }
 
 func (p *ProspectorLog) Run() {
-
 	logp.Debug("prospector", "Start next scan")
+
 	p.scan()
+	p.cleanupState()
+}
 
-	// Clear out files that disappeared and we've stopped harvesting
-	for file, lastinfo := range p.harvesterStats {
-		if lastinfo.Finished() && lastinfo.LastIteration < p.iteration {
-			delete(p.harvesterStats, file)
+// cleanupState cleans up the internal prospector state after each scan
+// Files which reached ignore_older are removed from the state as these states are not needed anymore
+func (p *ProspectorLog) cleanupState() {
```

Comment:
CleanupState should be in Prospector. Will move it to prospector.

Side note: yes the two are still partially overlapping, but much less now.